pyannote/database/registry.py

```diff
@@ -0,0 +1,473 @@
+#!/usr/bin/env python
+# encoding: utf-8
+
+# The MIT License (MIT)
+
+# Copyright (c) 2022 CNRS
+
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+# AUTHORS
+# Hervé BREDIN - http://herve.niderb.fr
+
+from enum import Enum
+import os
+from pathlib import Path
+from typing import Dict, List, Optional, Set, Text, Tuple, Type, Union
+import warnings
+
+from pyannote.database.protocol.protocol import Preprocessors, Protocol
+from .custom import create_protocol, get_init, get_custom_protocol_class_name
+from .database import Database
+import yaml
+
+
+class OverrideType(Enum):
+    OVERRIDE = 0  # replace existing
+    WARN_OVERRIDE = 1  # warn when replacing existing data, do it, and continue.
+    WARN_KEEP = 2  # warn when trying to replace existing data. Dont override it.
+    KEEP = 3  # never replace existing data
+
+
+class Registry:
+    """Stores the data from one (or multiple !) database.yml files."""
+
+    def __init__(self) -> None:
+        # Mapping of database.yml paths to their config in a dictionary
+        self.configs: Dict[Path, Dict] = dict()
+
+        # Content of the "Database" root item (=where to find file content)
+        self.sources: Dict[Text, List[Text]] = dict()
```

Comment:
Can you explain why `sources` attribute is needed (like I am five years old :))

Reply:
The only use of it is in `FileFinder`, which is called (`__call__`) with a `current_file:ProtocolFile` parameter, and wants to find its location (=filesystem path).
Currently, to do that the FileFinder can retrieve the template(s)/"sources" (eg `/home/data/{uri}.wav`) of current_file's database with `registry.sources[current_file.database]`. 
It then tries to find a file on the filesystem according to the acquired templates and the URI of the passed ProtocolFile.

`registry.sources` templates are already converted to absolute paths, because each database.yml may specify relative paths/templates.

It would be possible to remove `registry.sources` and replace it by a `def get_sources(database:str) -> List[Text]`,  in which case all database.yml configurations are iterated on and absolute paths recomputed on each call.
(Or the opposite, we could precompute the dictionary `Dict[DatabaseName, Dict[URI, Path]]` so that FileFinder just returns `registry[current_file.database][current_file.uri]`)

Reply:
> (Or the opposite, we could precompute the dictionary `Dict[DatabaseName, Dict[URI, Path]]` so that FileFinder just returns `registry[current_file.database][current_file.uri]`)

I'd rather not do that because it could take a veeeeeery long time for databases with huuuuuuuge number of files. That is the whole point of doing it lazily when accessing  `ProtocolFile` keys.

Reply:
To clarify the above comment.

When instantianting a protocol with `get_protocol(..., {"audio", FileFinder()})`, `FileFinder()(file)` is only called (and its result cached) when accessing `file["audio"]`. Same goes for `file["annotation"]` if the `annotation` key contains a `{uri}` placeholder. The point of this lazy loading mechanism is to avoid long initialization time when calling `get_protocol`.

Reply:
If registry.sources is removed, i think it'd be replace with :
```python
def get_sources(database:str) -> List[Text]
  sources = []
  for cfg in self.configs:
    if 'Databases' in cfg and database in cfg["Databases"]:
      sources += cfg["Databases"][database]
  return sources
```

Is this OK (to recompute all sources every call) ?
If I understand correctly FileFinder calls are cached, so should we ignore the extra computational cost it since it'll only happen once per file ?